### PR TITLE
CogView4 Control Block

### PR DESCRIPTION
**Pull request by zRzRzRzRzRzRzR**
What does this pull request do?

The purpose of this PR is to add a Control module to [CogView4](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/cogview4/pipeline_cogview4.py), which refers to the implementation of [Flux](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux_control.py).

Who can review?

@arrow

### CODE REVIEW

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+    ):
+        super().__init__()
+
+        self.register_modules(
+            tokenizer=tokenizer, text_encoder=text_encoder, vae=vae, transformer=transformer, scheduler=scheduler
+        )
+        self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1) if getattr(self, "vae", None) else 8
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+
+    def _get_glm_embeds(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:177]: add `#Copied` from here

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+            text_input_ids.to(self.text_encoder.model.device), output_hidden_states=True
+        ).hidden_states[-2]
+
+        prompt_embeds = prompt_embeds.to(dtype=dtype, device=device)
+        _, seq_len, _ = prompt_embeds.shape
+        prompt_embeds = prompt_embeds.repeat(1, num_images_per_prompt, 1)
+        prompt_embeds = prompt_embeds.view(batch_size * num_images_per_prompt, seq_len, -1)
+        return prompt_embeds
+
+    def encode_prompt(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:227]: same here

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+        image_seq_len = ((height // self.vae_scale_factor) * (width // self.vae_scale_factor)) // (
+            self.transformer.config.patch_size**2
+        )
+
+        timesteps = (
+            np.linspace(self.scheduler.config.num_train_timesteps, 1.0, num_inference_steps)
+            if timesteps is None
+            else np.array(timesteps)
+        )
+        timesteps = timesteps.astype(np.int64)
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:630]:
```suggestion
        timesteps = timesteps.astype(np.int64).astype(np.float32)
```

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+        sigmas = timesteps / self.scheduler.config.num_train_timesteps if sigmas is None else sigmas
+        mu = calculate_shift(
+            image_seq_len,
+            self.scheduler.config.get("base_image_seq_len", 256),
+            self.scheduler.config.get("base_shift", 0.25),
+            self.scheduler.config.get("max_shift", 0.75),
+        )
+        _, num_inference_steps = retrieve_timesteps(self.scheduler, num_inference_steps, device, sigmas=sigmas, mu=mu)
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:638]:
```suggestion
        timesteps, num_inference_steps = retrieve_timesteps(self.scheduler, num_inference_steps, device, timesteps, sigmas=sigmas, mu=mu)
```

we updated our scheduler to work with cogview4 - is there any reason we still cannot use the 1scheduler.set_timesteps1 to set timesteps?

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+        timesteps = torch.from_numpy(timesteps).to(device)
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:639]:
```suggestion
```

**Review by yiyixuxu — COMMENTED** [review]
thanks for the PR!
do we already have a checkpoint for cogview3 control lora or is this mainly to support training?

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py [hunk 6]
```diff
@@ -0,0 +1,711 @@
...
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```python
+        >>> import torch
+        >>> from diffusers import CogView4Pipeline
+
+        >>> pipe = CogView4Pipeline.from_pretrained("THUDM/CogView4-6B", torch_dtype=torch.bfloat16)
```
[the remark this answers is not in the document]
Comment (zRzRzRzRzRzRzR) [reply src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:48]: Now there is none, but we expect to train a control checkpoint, and we are in the process of training.

**Comment by zRzRzRzRzRzRzR** [comment]
> thanks for the PR! do we already have a checkpoint for cogview3 control lora or is this mainly to support training?

This is to support training, and furthermore, it supports Control training for the CogView4 model, not the CogView3 model.

File: src/diffusers/models/transformers/transformer_cogview4.py [hunk 7]
```diff
@@ -35,7 +36,7 @@ class CogView4PatchEmbed(nn.Module):
     def __init__(
         self,
         in_channels: int = 16,
-        hidden_size: int = 2560,
+        hidden_size: int = 4096,
```
[the remark this answers is not in the document]
Comment (zRzRzRzRzRzRzR) [reply src/diffusers/models/transformers/transformer_cogview4.py]: Already fixed.

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py [hunk 8]
```diff
@@ -0,0 +1,711 @@
...
+        >>> pipe.to("cuda")
+
+        >>> prompt = "A photo of an astronaut riding a horse on mars"
+        >>> image = pipe(prompt).images[0]
+        >>> image.save("output.png")
+        ```
+"""
+
+
+def calculate_shift(
```
[the remark this answers is not in the document]
Comment (zRzRzRzRzRzRzR) [reply src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:58]: I tried to add, it should be in this format.

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py (hunk 1 above)
Comment (zRzRzRzRzRzRzR) [reply src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:180]: Should be added to encoder_prompt or def _get_glm_embeds.

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py (hunk 3 above)
Comment (zRzRzRzRzRzRzR) [reply src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py]: done

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yiyixuxu) [reply src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:58]: then maybe you can update the one in cogview4 and then add a `#Copied from`?

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py (hunk 1 above)
Comment (yiyixuxu) [reply src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:180]: we should still add a `#Copied from` here, otherwise if we modify both `encode_prompt` and `_get_glm_embeds` from cogview4 pipeline, the `encode_prompt` method for cogview4 here will update but _get_glm_embeds` will not and output will not be correct

**Review by yiyixuxu — COMMENTED** [review]
I left some nits but looks good to me
do you want to wait to merge after you have a checkpoint? this way we could add proper doc & test too

**Comment by zRzRzRzRzRzRzR** [comment]
I think this is okay, we are still trying to train checkpoints.

**Comment by zRzRzRzRzRzRzR** [comment]
I left some nits but it looks good to me. Do you want to wait to merge after you have a checkpoint? This way, we could add proper documentation and tests too

I modified the method of encoding the prompt. Do I still need to add "copy from" in the fine-tuning of the downstream task? The prompt length is not variable, so we set it to a fixed length, and thus the padding method needs to be modified as well.

**Comment by yiyixuxu** [comment]
>I modified the method of encoding the prompt. Do I still need to add "copy from" in the fine-tuning of the downstream task? The prompt length is not variable, so we set it to a fixed length, and thus the padding method needs to be modified as well.

in that case, we don't need `#Copied from`!

**Comment by zRzRzRzRzRzRzR** [comment]
I have added a script for converting from the meagtron model of controlnet to the diffusers model.

File: src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py [hunk 9]
```diff
@@ -0,0 +1,713 @@
...
+
+        image_batch_size = image.shape[0]
+
+        if image_batch_size == 1:
+            repeat_by = batch_size
+        else:
+            # image batch size is the same as prompt batch size
+            repeat_by = num_images_per_prompt
+
+        image = image.repeat_interleave(repeat_by, dim=0)
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/cogview4/pipeline_cogview4_control.py:344]:
```suggestion
        image = image.repeat_interleave(repeat_by, dim=0, output_size=image.shape[0] * repeat_by)
```

Thanks to the recent find by hlky in #11030

**Review by a-r-r-o-w — APPROVED** [review]
Thanks for the amazing work and upcoming model! Just some minor changes remaining before we can merge

File: src/diffusers/models/transformers/transformer_cogview4.py [hunk 10]
```diff
@@ -156,6 +158,15 @@ def __call__(
...
         # 4. Attention
+        if attention_mask is not None:
+            text_attention_mask = attention_mask.float().to(query.device)
+            actual_text_seq_length = text_attention_mask.size(1)
+            new_attention_mask = torch.zeros((batch_size, text_seq_length + image_seq_length), device=query.device)
+            new_attention_mask[:, :actual_text_seq_length] = text_attention_mask
+            new_attention_mask = new_attention_mask.unsqueeze(2)
+            attention_mask_matrix = new_attention_mask @ new_attention_mask.transpose(1, 2)
+            attention_mask = (attention_mask_matrix > 0).unsqueeze(1).to(query.dtype)
+
```
[the remark this answers is not in the document]
Comment (zRzRzRzRzRzRzR) [reply src/diffusers/models/transformers/transformer_cogview4.py:169]: This part is for training, not for inference.

[the remark this answers is not in the document]
Comment (zRzRzRzRzRzRzR) [reply src/diffusers/models/transformers/transformer_cogview4.py:169]: If this part of the construction is not used, the corresponding attn mask is all 1s, which led to poor performance of the trained model in previous experiments.

File: src/diffusers/models/transformers/transformer_cogview4.py [hunk 11]
```diff
@@ -289,7 +304,7 @@ def forward(self, hidden_states: torch.Tensor) -> Tuple[torch.Tensor, torch.Tens
         return (freqs.cos(), freqs.sin())
 
 
-class CogView4Transformer2DModel(ModelMixin, ConfigMixin, PeftAdapterMixin):
+class CogView4Transformer2DModel(ModelMixin, ConfigMixin, PeftAdapterMixin, CacheMixin):
```
[the remark this answers is not in the document]
Comment (zRzRzRzRzRzRzR) [reply src/diffusers/models/transformers/transformer_cogview4.py:307]: I have increased the price for this part of the code, is this correct?

File: examples/cogview4-control/train_control_cogview4.py [hunk 12]
```diff
@@ -0,0 +1,1242 @@
...
+                    timestep=timesteps,
+                    original_size=original_size,
+                    target_size=target_size,
+                    crop_coords=crops_coords_top_left,
+                    return_dict=False,
+                    attention_mask=attention_mask,
+                )[0]
+                # these weighting schemes use a uniform timestep sampling
+                # and instead post-weight the loss
+                weighting = compute_loss_weighting_for_sd3(weighting_scheme=args.weighting_scheme, sigmas=sigmas)
```
[the remark this answers is not in the document]
Comment (zRzRzRzRzRzRzR) [reply examples/cogview4-control/train_control_cogview4.py:1131]: How do you think the convergence can be modified to be faster in this place?

**Comment by yiyixuxu** [comment]
@bot /style

**Comment by zRzRzRzRzRzRzR** [comment]
@bot /style

**Comment by yiyixuxu** [comment]
@bot /style
